### PR TITLE
AC-1018: Link Active Directory account to Keycloak user

### PR DESCRIPTION
**Pull request by mehranrashid**
### Description

Fixes [AC-1018](https://answerconsulting.jira.com/browse/AC-1018)

Allows users to register with AIDE using their Active Directory accounts. 

Newly created user:
image
image

Existing user with their AD account not linked:
image
image

Evidence in Keycloak UI for both:
image

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New unit tests added to cover the changes.
- [ ] New e2e tests added to cover the changes. (If it has been decided these will not be added with the PR, a seperate ticket **must** be created and linked in the ticket below before merge)
- [x] All tests passed locally.

### CODE REVIEW

File: src/common/envs/.env [hunk 1]
```diff
@@ -1,5 +1,8 @@
 PORT=5000
 
+ACTIVE_DIRECTORY_ENABLED=false
+EXTERNAL_IDENTITY_PROVIDER: 'active-directory'
```
Comment (samrooke) [inline src/common/envs/.env:4]: should this be an `=` instead of a `:`?
If so be sure to do the same for the other files.

Comment (mehranrashid) [reply src/common/envs/.env:4]: 🦅 👀

**Comment by woodheadio** [comment]
Noticed that you're using [firstname].[lastname] as the userName property you're passing up. I've actually managed to get the SAML configuration working so that it's the email address that's used for both, so we don't need to guess at the format they're using. Please could you use email address here instead?

**Comment by mehranrashid** [comment]
> Noticed that you're using [firstname].[lastname] as the userName property you're passing up. I've actually managed to get the SAML configuration working so that it's the email address that's used for both, so we don't need to guess at the format they're using. Please could you use email address here instead?

Changed userId and userName reference for the identity provider to be the email instead

File: src/modules/users/users.service.ts [hunk 2]
```diff
@@ -138,17 +165,77 @@ export class UsersService {
...
-        email: user.email,
-        firstName: user.firstName,
-        lastName: user.lastName,
-        enabled: user.enabled,
-      }),
+    const activeDirectoryEnabled = parseBoolean(
+      this.config.get('ACTIVE_DIRECTORY_ENABLED'),
+    );
+
+    if (!activeDirectoryEnabled) {
```
Comment (woodheadio) [inline src/modules/users/users.service.ts:172]: Can we  rename the feature flag. Sorry to be pedantic, but it isn't necessarily an Active Directory we're connecting to. Please could we rename to: `IDENTITY_PROVIDERS_ENABLED`

Comment (woodheadio) [reply src/modules/users/users.service.ts:172]: Will need to ripple through logging lines etc... sorry

File: src/modules/users/users.exceptions.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  readonly code: UserServiceExceptionCode;
+}
+
+export class UndefinedIdentityProviderException extends InternalServerErrorException {
+  constructor() {
+    super('identityProvider is undefined');
+  }
+}
+
+export class ActiveDirectoryProviderException extends InternalServerErrorException {
```
Comment (woodheadio) [inline src/modules/users/users.exceptions.ts:23]: `IdentityProviderNotFoundException`

File: src/modules/users/users.service.ts [hunk 4]
```diff
@@ -278,18 +365,88 @@ export class UsersService {
...
+    // Therefore, a workaround for getting the existing federated identity and then adding
+    // has been implemented.
+
+    // Get the identity provider
+    const activeDirectoryIdentityProvider:
+      | IdentityProviderRepresentation
+      | undefined = await this.adminService.performAction((realm, client) =>
+      client.identityProviders.findOne({
+        alias: this.config.get<string>('EXTERNAL_IDENTITY_PROVIDER'),
+        realm,
```
Comment (woodheadio) [inline src/modules/users/users.service.ts:427]: The update needs to change to reference a passed in Identity Provider, not a singular one from config

File: src/modules/users/users.service.ts [hunk 5]
```diff
@@ -278,18 +365,88 @@ export class UsersService {
...
+    userId: string,
+    user: EditUserDto,
+    identityProviderDetails: FederatedIdentityRepresentation,
+  ): Promise<User> {
+    // Method to update user and including the federatedIdentities property does not work.
+    // Therefore, a workaround for getting the existing federated identity and then adding
+    // has been implemented.
+
+    // Get the identity provider
+    const activeDirectoryIdentityProvider:
```
Comment (woodheadio) [inline src/modules/users/users.service.ts:422]: rename to `identityProvider`

File: src/modules/users/users.service.ts [hunk 6]
```diff
@@ -138,17 +165,77 @@ export class UsersService {
...
+        // Find method used as findOne requires user ID
+        client.users.find({
+          realm,
+          email: user.email,
+        }),
+    );
+
+    if (existingUser.length > 0) {
+      this.logger.debug(
+        'User exists in Keycloak. Checking if they already have their Active Directory account linked',
```
Comment (woodheadio) [inline src/modules/users/users.service.ts:210]: User exists in Keycloak. Checking if they already have their **identity provider** linked

File: src/modules/users/users.service.ts [hunk 7]
```diff
@@ -138,17 +165,77 @@ export class UsersService {
...
+      if (existingUser[0].federatedIdentities?.length > 0) {
+        this.logger.debug(
+          `User already has their account linked: ${existingUser[0].federatedIdentities}`,
+        );
+
+        return await this.getUser(existingUser[0].id);
+      }
+
+      this.logger.debug(
+        'Attempting to map the Active Directory details with the Keycloak account',
```
Comment (woodheadio) [inline src/modules/users/users.service.ts:222]: Attempting to map the identity provider with the Keycloak account

**Review by woodheadio — APPROVED** [review]

**Pull request merged**